protocols/identify/src/id_transport.rs

```diff
@@ -0,0 +1,237 @@
+        // This loop is here so that we can continue polling until we're ready.
+        loop {
+            // In order to satisfy the borrow checker, we extract the state and temporarily put
+            // `Poisoned` instead.
+            match mem::replace(&mut self.state, IdRetreiverState::Poisoned) {
+                IdRetreiverState::OpeningSubstream(mut opening, config) => {
+                    match opening.poll() {
+                        Ok(Async::Ready(Some(substream))) => {
+                            let upgrade = apply::apply(substream, config, self.endpoint);
+                            self.state = IdRetreiverState::NegotiatingIdentify(upgrade);
+                        },
+                        Ok(Async::Ready(None)) => {
+                            return Err(IoError::new(IoErrorKind::Other, "remote refused our identify attempt"));
+                        },
+                        Ok(Async::NotReady) => {
+                            self.state = IdRetreiverState::OpeningSubstream(opening, config);
+                            return Ok(Async::NotReady);
+                        },
+                        Err(err) => return Err(err),
+                    }
+                },
+                IdRetreiverState::NegotiatingIdentify(mut nego) => {
+                    match nego.poll() {
+                        Ok(Async::Ready(IdentifyOutput::RemoteInfo { info, .. })) => {
+                            self.state = IdRetreiverState::Finishing(info.public_key);
+                        },
+                        Ok(Async::Ready(IdentifyOutput::Sender { .. })) => {
+                            unreachable!("IdentifyOutput::Sender can never be the output from \
+                                          the dialing side");
+                        },
+                        Ok(Async::NotReady) => {
+                            self.state = IdRetreiverState::NegotiatingIdentify(nego);
+                            return Ok(Async::NotReady);
+                        },
+                        Err(err) => return Err(err),
+                    }
+                },
+                IdRetreiverState::Finishing(public_key) => {
+                    // Here is a tricky part: we need to get back the muxer in order to return
+                    // it, but it is in an `Arc`.
+                    let muxer = self.muxer.take()
+                        .expect("we only extract muxer when we are Finishing ; the Finishing \
+                                 state transitions into Poisoned ; therefore muxer is only \
+                                 extracted once ; qed");
+                    let unwrapped = Arc::try_unwrap(muxer).unwrap_or_else(|_| {
+                        panic!("we clone the Arc only to put it into substreams ; once in the \
+                                Finishing state, no substream or upgrade exists anymore ; \
+                                therefore there exists only one instance of the Arc ; qed")
```

Comment:
Why not removing the `Arc` and using `StreamMuxer::open_outbound` and `StreamMuxer::poll_outbound` instead? (with `TMuxer::Substream: AsyncRead + AsyncWrite ...` )

Reply:
The "problem" is that we use the `ConnectionUpgrade` trait, and that takes ownership of the substream, which thus requires holding an `Arc`.

Reply:
I must be missing something because the change looks rather good to me: https://gist.github.com/twittner/3bbc474ba9599880fe063d8e7f82a47d

Reply:
That's not correct though. `StreamMuxer::Substream` is an opaque type that isn't supposed to be implementing any trait.

Reply:
Ah yes, I got carried away by just following compiler suggestions and forgot about the semantics of `StreamMuxer`, also because yamux streams do implement AsyncRead and AsyncWrite.

Reply:
However it does not feel right if the API leads to unwrap-or-else-panic idioms.

Reply:
Well, this is a special case because we are at the level before the node handler comes into play.

Normally the code that handles a protocol only needs a substream and not the whole muxer.